egs/chime4/s5_1ch/local/chime4_train_lms.sh

```diff
@@ -0,0 +1,163 @@
+   | awk -v voc=$dir/vocab_5k.txt '
+   BEGIN{ while((getline<voc)>0) { invoc[$1]=1; }}
+   /^</{next}{
+     for (x=1;x<=NF;x++) {
+       w=toupper($x);
+       if (invoc[w]) { printf("%s ",w); } else { printf("<UNK> "); }
+     }
+     printf("\n");
+   }' | gzip -c > $dir/train.gz
+fi
+
+# get validation data from Chime4 dev set
+touch $dir/valid.gz
+if [ `du -k $dir/valid.gz | cut -f 1` -eq 68 ]; then
+  echo "Not getting validation data again [already exists]";
+else
+  echo "Collecting validation data from $chime4_data/data/transcriptions";
+  cut -d" " -f2- $chime4_data/data/transcriptions/dt05_real.trn_all \
+                 $chime4_data/data/transcriptions/dt05_simu.trn_all \
+      |gzip -c > $dir/valid.gz
+fi
+
+# train a large n-gram language model
+lm_suffix=${order}gkn_5k
+if [ -f $dir/lm_${lm_suffix}.arpa.gz ]; then
+  echo "A $order-gram language model aready exists and is not constructed again"
+  echo "To reconstruct, remove $dir/$dir/lm_${lm_suffix}.arpa.gz first"
+else
+  echo "Training a $order-gram language model"
+  ngram-count -text $dir/train.gz -order $order \
+              -vocab $dir/vocab_5k.txt -unk -map-unk "<UNK>" \
+              -gt2min 1 -gt3min 1 -gt4min 2 -gt5min 2 \
+              -interpolate -kndiscount \
+              -lm $dir/lm_${lm_suffix}.arpa.gz
+fi
+echo "Checking validation perplexity of $order-gram language model"
+ngram -order $order -ppl $dir/valid.gz -lm $dir/lm_${lm_suffix}.arpa.gz
+# e.g. 5-gram perplexity:
+# file data/local/local_lm/valid.txt: 3280 sentences, 54239 words, 3 OOVs
+# 0 zeroprobs, logprob= -96775.5 ppl= 48.1486 ppl1= 60.8611
+
+# Next, create the corresponding FST and lang_test_* directory.
+echo "Preparing language models for test"
+tmpdir=data/local/lm_tmp
+lexicon=data/local/lang_tmp/lexiconp.txt
+mkdir -p $tmpdir
+
+test=data/lang_test_${lm_suffix}
+mkdir -p $test
+for f in phones.txt words.txt phones.txt L.fst L_disambig.fst \
+   phones; do
+  cp -r data/lang/$f $test
+done
+gunzip -c $dir/lm_${lm_suffix}.arpa.gz | \
+ utils/find_arpa_oovs.pl $test/words.txt  > $tmpdir/oovs_${lm_suffix}.txt
+
+# grep -v '<s> <s>' because the LM seems to have some strange and useless
+# stuff in it with multiple <s>'s in the history.  Encountered some other similar
+# things in a LM from Geoff.  Removing all "illegal" combinations of <s> and </s>,
+# which are supposed to occur only at being/end of utt.  These can cause
+# determinization failures of CLG [ends up being epsilon cycles].
+gunzip -c $dir/lm_${lm_suffix}.arpa.gz | \
+  grep -v '<s> <s>' | \
+  grep -v '</s> <s>' | \
+  grep -v '</s> </s>' | \
+  arpa2fst - | fstprint | \
+  utils/remove_oovs.pl $tmpdir/oovs_${lm_suffix}.txt | \
+  utils/eps2disambig.pl | utils/s2eps.pl | fstcompile --isymbols=$test/words.txt \
+    --osymbols=$test/words.txt  --keep_isymbols=false --keep_osymbols=false | \
+   fstrmepsilon | fstarcsort --sort_type=ilabel > $test/G.fst
+fstisstochastic $test/G.fst
+# The output is like:
```

Comment:
I would suggest to replace this by a script (arpa2G.sh which can be found in some other recipes. Not sure if it would cover correctly your use-case, but most probably yes. Also, we have a newly written arpa2fst, which IIRC already does some of these things